### PR TITLE
[cxx-interop] Add ability to specify protocol conformance on C++ side.

### PR DESCRIPTION
**Pull request by zoecarver**

### CODE REVIEW

**Comment by zoecarver** [comment]
@swift-ci please test

**Review by ktoso — APPROVED** [review]
Really awesome, thank you @zoecarver :)

**Comment by hyp** [comment]
We should error out if the protocol lookup fails to catch typos .

**Comment by hyp** [comment]
we should also error out on ambiguous lookups, no?

**Comment by zoecarver** [comment]
@swift-ci please test

File: lib/AST/ProtocolConformance.cpp [hunk 1]
```diff
@@ -1052,49 +1052,48 @@ void NominalTypeDecl::prepareConformanceTable() const {
...
 
-  auto addSynthesized = [&](KnownProtocolKind kind) {
-    if (auto *proto = getASTContext().getProtocol(kind)) {
-      if (protocols.count(proto) == 0) {
-        ConformanceTable->addSynthesizedConformance(
-            mutableThis, proto, mutableThis);
-        protocols.insert(proto);
-      }
+  auto addSynthesized = [&](ProtocolDecl *proto) {
+    llvm::dbgs() << "S\n";
```
Comment (hyp) [inline lib/AST/ProtocolConformance.cpp:1056]: Debug output, pls rm -rf.

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci please test

**Review by egorzhdan — COMMENTED** [review]
This is a great improvement but I'm a bit worried that importing a C++ module can now trigger loading of an unrelated Swift module, which might run into cyclic dependencies. 

Could you please add a test that has something like this:
1. a C++ module `A` that contains a decl with `conforms_to:X` attr
2. a Swift module `B` that imports `A`
3. a Swift module `C` that imports `B` and contains `protocol X`

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
> This is a great improvement but I'm a bit worried that importing a C++ module can now trigger loading of an unrelated Swift module, which might run into cyclic dependencies.
> 
> Could you please add a test that has something like this:
> 
> 1. a C++ module `A` that contains a decl with `conforms_to:X` attr
> 2. a Swift module `B` that imports `A`
> 3. a Swift module `C` that imports `B` and contains `protocol X`

Yeah this is a good point. Regardless, it kind of sucks that we have to load all the modules and potentially have ambiguities. What if we just required people to specify the module in the attribute string?

**Comment by ktoso** [comment]
> What if we just required people to specify the module in the attribute string?

I'd be more than happy with that; requiring to be precise here sounds good tbh.

**Comment by zoecarver** [comment]
I will address the review comments post-commit. I think we should require the user to specify both the module and the protocol.

**Pull request merged**